# Data Analyst Practice Case - Iykra Week 3
This is practice case in week 3 at Data Science Fellowship IYKRA. I will answer some questions related hotel dataset.  

In [1]:
import pandas as pd
import datetime

In [2]:
df = pd.read_csv("Ritz_Jager_Data.csv")

In [3]:
df.head()

hotel_type  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  deposit_type  \
0                        0                     0       2  ...    No Deposit   
1                        0                     0       2  ...    No Deposit   
2                        0                     1       1  ...    No Deposit   
3                        0                     1       1  ...    No Deposit   
4                        0                     2       2  ...    No Deposit   

   agent company days_in_waiting_list customer_type   adr  \
0    NaN     NaN                    0     Transient   0.0   
1    NaN     NaN                    0     Transient   0.0   
2    NaN     NaN                    0     Transient  75.0   
3  304.0     NaN                    0     Transient  75.0   
4  240.0     NaN                    0     Transient  98.0   

   required_car_parking_spaces  total_of_special_requests  reservation_status  \
0                            0                          0           Check-Out   
1                            0                          0           Check-Out   
2                            0                          0           Check-Out   
3                            0                          0           Check-Out   
4                            0                          1           Check-Out   

  reservation_status_date  
0                1/7/2015  
1                1/7/2015  
2                2/7/2015  
3                2/7/2015  
4                3/7/2015  

[5 rows x 32 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel_type                      119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal_type       

In [5]:
df.isna().sum()

hotel_type                             0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal_type                              0
country_origin                       488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

**There are missing value on columns country_origin, agent and comapny**

## Question 1: These are the countries where the guests come from

**First, we filter the quests with comfirm the booking**

In [13]:
df_yes = df[df['is_canceled']==0]

**So, these are the top 5 countries where the quests come from**

In [14]:
df_yes['country_origin'].value_counts().head()

PRT    21071
GBR     9676
FRA     8481
ESP     6391
DEU     6069
Name: country_origin, dtype: int64

## Question 2: These are the price of a room per night

**We have information about average cost per night per quest in column 'adr', so we use this value to calculate the price per room per night.**

In [16]:
df_yes1 = df_yes[['hotel_type','arrival_date_year','arrival_date_month','arrival_date_day_of_month','market_segment','assigned_room_type', 'adr','reservation_status','reservation_status_date']]
df_pos = df_yes1[df_yes1['adr']>0]

**We create the pivot table group by hotel type and room type and calculate the average of 'adr' to find the price of a room per night.**

In [17]:
pd.pivot_table(df_pos, index = ['hotel_type', 'assigned_room_type'], values = 'adr', aggfunc = 'mean')

adr
hotel_type   assigned_room_type            
City Hotel   A                    99.078326
             B                    97.518135
             C                   110.417879
             D                   117.913473
             E                   142.225000
             F                   178.171620
             G                   196.841696
             K                   102.107391
Resort Hotel A                    81.309890
             B                   105.815517
             C                   108.913320
             D                    82.425891
             E                   102.462598
             F                   120.389556
             G                   153.179443
             H                   162.198575
             I                    90.872089

## Question 3: The price of a room per night vary over year.

**We create the pivot table group by year, hotel type and room type and calculate the average of 'adr' to find the price of a room per night vary over year.**

In [11]:
pd.pivot_table(df_pos, index = ['arrival_date_year','hotel_type', 'assigned_room_type'], values = 'adr', aggfunc = 'mean')

adr
arrival_date_year hotel_type   assigned_room_type            
2015              City Hotel   A                    89.216429
                               B                    88.834053
                               C                    82.378571
                               D                    91.291770
                               E                   110.600417
                               F                   150.325351
                               G                   131.473750
                               K                    74.800000
                  Resort Hotel A                    79.281096
                               B                    97.998065
                               C                   110.028832
                               D                    83.851389
                               E                   104.439456
                               F                   120.154128
                               G                   137.662902
                               H                   142.850110
                               I                    46.695000
2016              City Hotel   A                    96.957262
                               B                    99.279515
                               C                   114.189194
                               D                   117.370879
                               E                   133.552214
                               F                   173.393545
                               G                   186.484924
                               K                    99.471216
                  Resort Hotel A                    76.908181
                               B                   103.178548
                               C                   100.329351
                               D                    74.365909
                               E                    93.472482
                               F                   111.210935
                               G                   148.271200
                               H                   158.079535
                               I                    77.395591
2017              City Hotel   A                   106.445907
                               B                   110.950169
                               C                   109.821905
                               D                   131.685257
                               E                   164.636312
                               F                   196.270224
                               G                   222.434545
                               K                   106.648361
                  Resort Hotel A                    89.028942
                               B                   113.620000
                               C                   119.667133
                               D                    93.916076
                               E                   114.642690
                               F                   133.781389
                               G                   171.516226
                               H                   180.704196
                               I                   114.315082

## Question 4: The busiest month.

**We create the pivot table group by year, month and calculate the count of 'adr' and sort by ascending to find the busiest month.**

In [12]:
pd.pivot_table(df_pos, index = ['arrival_date_year', 'arrival_date_month'], values = 'adr', aggfunc = 'count').sort_values(by='adr', ascending = False)

adr
arrival_date_year arrival_date_month      
2016              October             3601
2017              May                 3492
2016              May                 3460
                  September           3324
                  April               3294
2017              July                3290
2016              March               3272
2017              March               3232
2016              August              3199
2017              June                3161
                  April               3155
2016              June                3135
2015              October             3119
2017              August              3084
2016              July                3018
2015              September           2935
2017              February            2751
2016              November            2738
                  February            2494
                  December            2376
2017              January             2373
2015              August              2200
                  December            1830
                  November            1793
2016              January             1633
2015              July                1460

## Question 5: Stay duration in hotel.

**To find the stay duration in hotel, we have to find the difference of arrive date and leave date in hotel and fill in this value to new column (create new column).** 

In [22]:
df_pos['arrive'] = df_pos['arrival_date_day_of_month'].astype(str)+'/'+df_pos['arrival_date_month'].astype(str)+'/'+df_pos['arrival_date_year'].astype(str)
df_pos['arrive_date'] = pd.to_datetime(df_pos['arrive'], format='%d/%B/%Y')
df_pos['leave_date'] = pd.to_datetime(df_pos['reservation_status_date'], format='%d/%m/%Y')
df_pos['dur_stay'] = df_pos['leave_date'] - df_pos['arrive_date']


C:\Users\Azmul Azmi\anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
C:\Users\Azmul Azmi\anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
C:\Users\Azmul Azmi\anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See

In [23]:
df_pos['dur_stay'].astype('timedelta64[D]').value_counts().head()

3.0    15608
1.0    15279
2.0    15156
4.0    10974
7.0     5662
Name: dur_stay, dtype: int64

In [15]:
df_pos['dur_stay'].astype('timedelta64[D]').mean()

3.433470899903295

**So, average duration stay is 3 days - 4 days.**

## Question 6: Bookings by market segment.

**We create the pivot table group by market segment and calculate the count of reservation status to find the bookings by market segment.**

In [37]:
pd.pivot_table(df_yes1, index = 'market_segment', values = 'reservation_status', aggfunc = 'count').sort_values(by='reservation_status', ascending = False)

reservation_status
market_segment                    
Online TA                    35738
Offline TA/TO                15908
Direct                       10672
Groups                        7714
Corporate                     4303
Complementary                  646
Aviation                       185

## Question 7: Count of bookings that canceled.

**We filter the column 'is canceled' with value =1.** 

In [26]:
df_no = df[df['is_canceled']==1]
df_no['is_canceled'].count()

44224

**So, there are 44224 bookings that canceled.**

## Question 8: The month of highest number of canceled.

**We create the pivot table group by year and month and calculate the count of cancel status to find the The month of highest number of canceled.**

In [27]:
pd.pivot_table(df_no, index = ['arrival_date_year', 'arrival_date_month'], values = 'is_canceled', aggfunc = 'count').sort_values(by='is_canceled', ascending=False)

is_canceled
arrival_date_year arrival_date_month             
2017              May                        2762
2016              October                    2514
2017              April                      2463
                  June                       2439
2016              June                       2096
2015              September                  2094
2016              April                      2061
                  September                  2022
2017              July                       1984
2016              May                        1915
                  August                     1825
2017              August                     1816
2015              October                    1732
2017              March                      1672
2016              November                   1636
2015              August                     1598
2016              July                       1499
                  March                      1477
                  December                   1398
2017              February                   1359
2016              February                   1337
2015              July                       1259
2017              January                    1250
2015              December                    973
2016              January                     557
2015              November                    486